zzre/game/systems/dialog/DialogTrading.cs

```diff
@@ -0,0 +1,245 @@
+﻿using System;
+using System.Linq;
+using System.Numerics;
+using System.Collections.Generic;
+using zzio;
+using zzio.db;
+
+namespace zzre.game.systems;
+
+public partial class DialogTrading : ui.BaseScreen<components.ui.DialogTrading, messages.DialogTrading>
+{
+    private static readonly components.ui.ElementId IDExit = new(1000);
+    private static readonly components.ui.ElementId IDYes = new(1001);
+    private static readonly components.ui.ElementId IDNo = new(1002);
+
+    private readonly MappedDB db;
+    private readonly IDisposable resetUISubscription;
+
+    public DialogTrading(ITagContainer diContainer) : base(diContainer, BlockFlags.None)
+    {
+        db = diContainer.GetTag<MappedDB>();
+
+        resetUISubscription = World.Subscribe<messages.DialogResetUI>(HandleResetUI);
+        OnElementDown += HandleElementDown;
+    }
+
+    public override void Dispose()
+    {
+        base.Dispose();
+        resetUISubscription.Dispose();
+    }
+
+    private void HandleResetUI(in messages.DialogResetUI message)
+    {
+        foreach (var entity in Set.GetEntities())
+            entity.Dispose();
+    }
+
+    protected override void HandleOpen(in messages.DialogTrading message)
+    {
+        message.DialogEntity.Set(components.DialogState.Talk);
+
+        World.Publish(new messages.DialogResetUI(message.DialogEntity));
+        var uiEntity = World.CreateEntity();
+        uiEntity.Set(new components.Parent(message.DialogEntity));
+
+        uiEntity.Set(new components.ui.DialogTrading(
+            message.DialogEntity,
+            db.GetItem(message.CardUID),
+            message.CardTrades,
+            new()
+        ));
+        ref var trading = ref uiEntity.Get<components.ui.DialogTrading>();
+
+        trading.Profile = CreatePrimary(uiEntity, trading);
+    }
+
+    private DefaultEcs.Entity CreatePrimary(DefaultEcs.Entity parent, components.ui.DialogTrading trading)
+    {
+        var entity = World.CreateEntity();
+        entity.Set(new components.Parent(parent));
+
+        preload.CreateDialogBackground(entity, animateOverlay: false, out var bgRect);
+        CreateTopbar(entity, trading.Currency);
+        var trades = trading.CardTrades.ToArray();
```

Comment:
you do not have to convert `CardTrades` to an array, just use `Count` instead of `Length` and everything else stays the same.